pilot/pkg/model/xds_cache.go

```diff
@@ -130,15 +129,7 @@ type XdsCache interface {
 	Add(entry XdsCacheEntry, token CacheToken, value *discovery.Resource)
 	// Get retrieves the cached value if it exists. The boolean indicates
 	// whether the entry exists in the cache.
```

Comment:
This looks like it has the opportunity to be a big improvement - right now the cache is definitely very complex.

One concern I have - is the push request start time sufficient? For example, the EndpointShards may change during the middle of a push? For example:

```
|Push Request| Endpoint State | Thread 1            | Thread 2            |
|------------|----------------|---------------------|---------------------|
| Push 1     | v1             |                     |                     |
|            | v1             |                     |                     |
|            |                | Start generating v1 |                     |
|            | v2             |                     |                     |
|            |                |                     | Start generating v2 |
|            |                |                     | Write v2, token=1   |
|            |                | Write v1, token=1   |                     |
```

I think in this case we end up with endpoint state `v1` since the write isn't invalidated. Both writes use the same token but have different state

Reply:
In this case the v2 push.StartTime > v1 push.StartTime, so the `Write V1` will be stopped by 

```
	if token == 0 || token < l.token {
		return
	}

```

Reply:
l.token = v2 push.StartTime, 
token =   v1 push.StartTime

Reply:
In my example they are in the same push. The endpoint state (endpoint shards) can change during the middle of a push

Reply:
v1 and v2 event enqueued to pushChannel and dequeued from it is in order, right? 

And then they will be set StartTime (v1<v2)and  enqueued to pushQueue, even later v2 is processed (let's say generate eds)before v1, it will end up caching v2.

Reply:
> The endpoint state (endpoint shards) can change during the middle of a push

It does not influence the correctness, for every ep shard change, it will trigger a new round of eds push at least. The xds cache will then be renewed too.

Reply:
I think we may be mixing up "Push version" (ie a PushRequest) which has the start time, and the "endpoint version". We do not literally have an "endpoint version" (although we could easily add one: see https://github.com/istio/istio/pull/33620/files#diff-c6215f1e2a80c0d2afe120f4616972dfb339bcb51e8815cc5e171750bc11ae18R154) in the code, but there is a logical version that we can consider to be incremented each time edsUpdate occurs.

In the example where I think things may not work above, there is only every a single push.

Long sequence of events:
* Service Z scales from 0 -> 1. We consider this "Endpoint version v1". This event triggers a pushChannel update
* Push debounce completes (from "endpoint version v1" event), we start a push with StartTime=1
* For proxy-1, we start eds generate. When we fetch the endpoint shards, we get the endpoint version v1
* Service Z scales from 1 -> 2. We consider this "Endpoint version v2". This event triggers a pushChannel update
* For proxy-2, we start eds generate. Note: this is still from the push with StartTime=1
* proxy-2 completes EDS generation. It used "endpoint version v2" to compute this, so we have 2 endpoints. We write this to the cache. _the write succeeds since the token is not less than the current token (both are 1 from push with StartTime=1)_
* proxy-1 completes EDS generation. It used "endpoint version v1" to compute this, so we have 1 endpoint. We write this to the cache. The write succeeds since it was the first write, so no token issue. Token=1 (from StartTime). **BUG: we went backwards**
* Push debounce completes (from "endpoint version v2" event), we start a push with StartTime=2. This clears the cache.

Reply:
>proxy-1 completes EDS generation. It used "endpoint version v1" to compute this, so we have 1 endpoint. We write this to the cache. The write succeeds since it was the first write, so no token issue. Token=1 (from StartTime).

You are right, let me think about it

Reply:
It seems tricky, I think https://github.com/istio/istio/blob/4477ee5d75e64fcc0b7e890484278fa535ba4907/pilot/pkg/xds/eds.go#L152-L159 and here https://github.com/istio/istio/blame/4477ee5d75e64fcc0b7e890484278fa535ba4907/pilot/pkg/xds/eds.go#L191 are both for the eds cache.

How about adding a flag in xDSCache to control whether to allow `Add` with the same token the cache has? When clear the cache, we mark the flag = true, and then set a data like eds with the same token the cache already has, prevent it.  only allow cache with a newer token.

Reply:
Hmm.. I thought the goal of this PR was to avoid all of these places where we have to clear the cache under locks because its extremely complex (which I agree with). If we do it this way, then we still have the same issue so I am not sure its much simpler?

Reply:
With it we can remove clear cache in different places, and avoid Get before Add.

And likely the above case https://github.com/istio/istio/pull/33611#discussion_r660251243 can be resolved

Reply:
Only do this once in `dropCacheForRequest` is enough.

Reply:
An alternative way is to set a newer token when calling clear cache, and directly reject cache with elder token.

Reply:
Changed to this